client/app/telecom/telephony/line/phone/programmableKeys/telecom-telephony-line-phone-programmableKeys.controller.js

```diff
@@ -83,6 +83,84 @@ angular.module("managerApp").controller("TelecomTelephonyLinePhoneProgammableKey
         });
     };
 
+    /* ===========================
+    =            BULK            =
+    ============================ */
+
+    self.bulkDatas = {
+        infos: {
+            name: "functionKeys",
+            actions: [{
+                name: ""
+            }]
+        }
+    };
+
+    self.getBulkParams = function () {
+        self.bulkDatas.infos.actions = self.buildBulkActions();
+    };
+
+    self.buildBulkActions = function () {
+        return _.map(self.functionKeys.raw, function (key) {
+            return {
+                name: "functionKey",
+                route: `/telephony/{billingAccount}/line/{serviceName}/phone/functionKey/${key.keyNum}`,
+                method: "PUT",
+                params: {
+                    "function": key.function,
+                    parameter: key.parameter
+                }
+            };
+        });
+    };
+
+    self.filterServices = function (services) {
+        var filteredServices = _.filter(services, function (service) {
+            return ["sip", "mgcp"].indexOf(service.featureType) > -1;
+        });
+
+        return voipLinePhoneFunction.fetchAll().then(function (voipLinePhoneFunctions) {
+            return $q.when(_.filter(filteredServices, function (service) {
```

Comment:
Why do return a promise ? You can directly return the result of `_.filter`. The promise that `filterServices` callback need is `voipLinePhoneFunction.fetchAll()` ;-)